src/graph_compute/chromatic_scheduling.h

```diff
@@ -0,0 +1,105 @@
+#ifndef CHROMATIC_SCHEDULING_H_
+#define CHROMATIC_SCHEDULING_H_
+
+#if D1_CHROM
+
+#include <algorithm>
+#include "./common.h"
+
+struct scheddata_t {
+  vid_t cntColors;
+  vid_t * cntNodesPerColor;
+  vid_t * nodesByColor;
+};
+typedef struct scheddata_t scheddata_t;
+
+struct sched_t {
+  vid_t color;
+};
+typedef struct sched_t sched_t;
+
+#include "./update_function.h"
+
+static inline vid_t colorGraph(vertex_t * const nodes,
+                              const vid_t cntNodes) {
+  bool * colors = new (std::nothrow) bool[cntNodes+1];
+  vid_t maxColor = 0;
+  for (vid_t v = 0; v < cntNodes; v++) {
+    for (vid_t edge = 0; edge <= nodes[v].cntEdges; edge++) {
+      colors[edge] = false;
+    }
+    for (vid_t edge = 0; edge < nodes[v].cntEdges; edge++) {
+      if (nodes[v].edges[edge] < v) {
+        colors[edge] = true;
+      }
+    }
+    static const vid_t MAX_COLOR = cntNodes + 1;
+    nodes[v].sched.color = MAX_COLOR;
+    for (vid_t edge = 0; edge <= nodes[v].cntEdges; edge++) {
+      //  color not yet taken
+      if (colors[edge] == false) {
+        if (nodes[v].sched.color == MAX_COLOR) {
+          nodes[v].sched.color = edge;
+          maxColor = std::max(maxColor, edge);
+        }
+      }
+    }
+  }
+  return maxColor;
+}
+
+static inline void init_scheduling(vertex_t * const nodes,
+                                   const vid_t cntNodes,
+                                   scheddata_t * const scheddata) {
+  scheddata->cntColors = colorGraph(nodes, cntNodes);
+  scheddata->cntNodesPerColor = new (std::nothrow) vid_t[scheddata->cntColors + 1]();
+  scheddata->nodesByColor = new (std::nothrow) vid_t[cntNodes];
+  for (vid_t v = 0; v < cntNodes; v++) {
+    scheddata->cntNodesPerColor[nodes[v].sched.color + 1]++;
```

Comment:
For convenience of calculating indices into nodesByColor, it's nice to take differences of the cumulative count array.  This is made easier by counting from 1 and having the zero entry equal 0.